src/common/system_logger.h

```diff
@@ -0,0 +1,91 @@
+#ifndef GFS_COMMON_SYSTEM_LOGGER_H_
+#define GFS_COMMON_SYSTEM_LOGGER_H_
+
+#include <glog/logging.h>
+#include <glog/stl_logging.h>
+
+#include <string>
+
+// The system_logger wraps the logging library, which in this case is Google
+// glog. The reason for having a wrapper is to ensure that we can easily change
+// the underlying logging library if needed without having to change the
+// system_logger clients code. Also this helps us to control the functionalities
+// the underlying library that we want to expose.
+
+namespace gfs {
+
+namespace common {
+
+// Singleton class for managing for configuring how we log. The system logger
+// must be intialized in the application (ideally once) before using any of the
+// logging macros. Please see examples/system_logger_main.cc on how to use.
+// We currently only log to console.
+class SystemLogger {
+ public:
+  static SystemLogger& GetInstance() {
+    static SystemLogger instance;
+
+    return instance;
+  }
+
+  // To prevent copying the SystemLogger
+  SystemLogger(const SystemLogger&) = delete;
+
+  void operator=(const SystemLogger&) = delete;
+
+  // This initializes the logging framework and does the necessary
+  // configurations. Currently logging to console only. Must be called before
+  // using any of the logging macros. This isn't thread safe and must be done
+  // from a single thread.
+  void Initialize(const std::string& program_name);
+
+ private:
+  bool is_initialized_;
+
+  SystemLogger() : is_initialized_(false) {}
+};
+
+}  // namespace common
+}  // namespace gfs
+
+// The macros you should use for logging. only use the SYS_* macros. Please see
+// examples/system_logger_main.cc on how to use.
+//
+// SEVERITY LEVELS: You can specify one of the following severity levels (in
+// increasing order of severity): INFO, WARNING, ERROR, and FATAL. Logging a
+// FATAL message terminates the program (after the message is logged). Note that
+// messages of a given severity are logged not only in the logfile for that
+// severity, but also in all logfiles of lower severity. E.g., a message of
+// severity FATAL will be logged to the logfiles of severity FATAL, ERROR,
+// WARNING, and INFO.
+// Further reading about glog:
+// https://github.com/google/glog/blob/master/doc/glog.html
+
+// Log with the specified severity.
+#define SYS_LOG(severity) LOG(severity)
```

Comment:
Could you please also clarify whether these logging MACROS are thread safe or not? The server processes are inherently multi-threading so it would be good if this can be thread-safe, otherwise, we may need to use locks to serialize logging.

Reply:
Great question, yes these macros are thread safe and we don't have to synchronize. glog does the underlying synchronization for us. Will add comment for that.

Reply:
Great, sounds good to me